pkg/controller/syncmachineset/syncmachineset_controller.go

```diff
@@ -0,0 +1,563 @@
+// remote cluster API
+type Controller struct {
+	client     coclient.Interface
+	kubeClient kubeclientset.Interface
+
+	// To allow injection of syncMachineSet for testing.
+	syncHandler func(queuedMS queuedMachineSet) error
+
+	// Used for unit testing
+	enqueueMachineSet func(machineSet *cov1.MachineSet)
+
+	buildClients func(cluster *cov1.MachineSet) (clusterapiclient.Interface, kubeclientset.Interface, error)
+
+	// machineSetsLister is able to list/get machine sets and is
+	// populated by the shared informer passed to NewController.
+	machineSetsLister lister.MachineSetLister
+	// machineSetsSynced returns true if the machine set shared
+	// informer has been synced at least once.  Added as a member
+	// to the struct to allow injection for testing.
+	machineSetsSynced cache.InformerSynced
+
+	// clusterLister is able to list/get clusters and is populated
+	// by the shared informer passed to NewController.
+	clusterLister lister.ClusterLister
+	// clusterSynced returns true if the cluster shared informer
+	// has been synced at least once. Added as a member to the
+	// struct to allow injection for testing.
+	clusterSynced cache.InformerSynced
+
+	// Machines that need to be synced
+	queue workqueue.RateLimitingInterface
+
+	logger log.FieldLogger
+}
+
+// queuedMachineSet represents data we need to process a machine set in the queue. This includes
+// data we might need to process deleted objects we can no longer lookup.
+type queuedMachineSet struct {
+	key         string
+	clusterName string
+}
+
+func (c *Controller) addMachineSet(obj interface{}) {
+	machineSet := obj.(*cov1.MachineSet)
+	logging.WithMachineSet(c.logger, machineSet).Debugf("Adding machine set")
+	c.enqueueMachineSet(machineSet)
+}
+
+func (c *Controller) updateMachineSet(old, cur interface{}) {
+	machineSet := cur.(*cov1.MachineSet)
+	logging.WithMachineSet(c.logger, machineSet).Debugf("Updating machine set")
+	c.enqueueMachineSet(machineSet)
+}
+
+// Run runs c; will not return until stopCh is closed. workers determines how
+// many machine sets will be handled in parallel.
+func (c *Controller) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	c.logger.Infof("Starting syncmachineset controller")
+	defer c.logger.Infof("Shutting down syncmachineset controller")
+
+	if !controller.WaitForCacheSync("syncmachineset", stopCh, c.machineSetsSynced) {
+		c.logger.Errorf("could not sync caches for syncmachineset controller")
+		return
+	}
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(c.worker, time.Second, stopCh)
+	}
+
+	<-stopCh
+}
+
+func (c *Controller) enqueue(machineSet *cov1.MachineSet) {
+	key, err := controller.KeyFunc(machineSet)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("couldn't get key for object %#v: %v", machineSet, err))
+		return
+	}
+
+	c.queue.Add(queuedMachineSet{
+		key:         key,
+		clusterName: machineSet.Labels[controller.ClusterNameLabel],
```

Comment:
Why are we verifying that the label exists in `enqueueAfter` but not in `enqueue`?